File: euler12.cpp
#include <iostream>
#include <vector>

inline int nth_triangle(int n) {
    return n * (n+1) / 2;
}

int count_divisors(int number) {
    std::vector<int> divisor_test(number/2, 1);
    std::size_t divisor_count = 0;
    std::size_t i = 0;
    while(i < divisor_test.size()) {
        if(divisor_test[i] == 1) {
            int divisor = i + 1;
            if(number % divisor == 0)
                divisor_count++;
            else {
                int non_divisor = divisor;
                while(non_divisor < number/2 + 1) {
                    divisor_test[non_divisor - 1] = 0;
                    non_divisor += divisor;
                }
            }
        }
        i++;
    }
    return divisor_count + 1;
}

int triangle_over_x_divisors(int x) {
    int divisor_count = 0;
    int triangle_index = 1;
    int triangle = 0;
    while(divisor_count <= x) {
        // triangle = nth_triangle(triangle_index);
        triangle += triangle_index;
        divisor_count = count_divisors(triangle);
        std::cout << triangle << ' ' << divisor_count << std::endl;
        triangle_index++;
    }
    return triangle;
}
int main(void) {
    std::cout << triangle_over_x_divisors(500) << std::endl;
    return 0;
}

File: euler12.py
from itertools import count

def nth_triangle(n):
    return n * (n+1) // 2

def next_triangle():
    for n in count(1):
        yield nth_triangle(n)
    
def count_divisors(number):
    divisor_test = [1] * int(number ** 0.5)
    divisor_count = 0
    i = 0
    while i < len(divisor_test):
        if divisor_test[i] == 1:
            # test divisibility by i+1 (0th index corresponds to 1 and so on)
            divisor = i + 1
            if number % divisor == 0:
                divisor_count += 2
            else:
                for non_divisor in range(divisor, len(divisor_test), divisor):
                    divisor_test[non_divisor - 1] = 0
        i += 1
    return divisor_count + 1

def triangle_over_x_divisors(x):
    for triangle in next_triangle():
        a = count_divisors(triangle)
        if a > x:
            return triangle

if __name__ == '__main__':
    print(triangle_over_x_divisors(500))
